### ISSUE TITLE
NILMTK Disagreggation algorithms

### ISSUE THREAD

**Issue by psychdragon** [nilmtk/nilmtk#424]
I have been using your toolkit for a while(NILMTK 0.2) and managed to write simple scripts just to compare the built in algorithms(CO and FHMM) with different datasets and test/train ratios. I have to say i am happy i stumbled upon it when i did.

I have a few questions though.
1. does it already have built in cross validation tools?
2. on top of the f-score, could i get confusion matrices displayed as well?
3. I'd also like to include some additional algorithms(sparse coding/SVN maybe). any guides on how I can do this? code templates maybe?(am quite new to python, but i should be able to hack my way around)

**Comment by nipunreddevil** [comment]
Hi @psychdragon 
Some really good questions there. I'll answer them briefly here. I think these questions can form a part of our cookbook.

> does it already have built in cross validation tools?

cross-validation for time series data can be a bit different from traditional cross validation. For instance, we can't randomly pick samples to be in train and test. We need to pick them up in sequence. We currently have an implementation which divides the data set into train and test. Both train and test are sequential data. 

Sample call

```python

elec = ds.buildings[b_id].elec
# Dividing train, test
train = DataSet(ds_path)
test = DataSet(ds_path)
split_point = elec.train_test_split(train_fraction=train_fraction).date()
train.set_window(end=split_point)
test.set_window(start=split_point)
train_elec = train.buildings[b_id].elec
test_elec = test.buildings[b_id].elec
```


>on top of the f-score, could i get confusion matrices displayed as well?

I think one would typically use confusion matrices when one is dealing with event based disaggregation. For each event you see what the actual event was and what the predicted event is. We currently don't have anything specialised to this effect. It must be noted that metrics of these sort need to be handled carefully. For instance, your algorithm could be off about 2 seconds from the actual event. So, this also needs to be taken into consideration. Having said that, I think it should be fairly easy to convert power to events.


>I'd also like to include some additional algorithms(sparse coding/SVN maybe). any guides on how I can do this? code templates maybe?(am quite new to python, but i should be able to hack my way around)

I think the best way to do so by using elementary Pandas data structures. As a first step, I'd recommend starting with a Pandas data frame for training where each column represents a different appliance. A similar data frame is passed for testing. The predicted data frame (after applying your algo) matches in structure to the passed data frame. Once this is done, it should be fairly easy to port your algo into nilmtk.

Hope this helps. Feel free to ask more.

**Comment by psychdragon** [comment]
My thanks for the explanation. it does help somewhat. I do have some other questions though, being new to the entire thing and using matlibplot etc.

one of it would be, if i wanted to plot out a selection of appliances together instead of the entire collection, how do i do that?
right now there is elec.plot() for all the sensors and elec[meternumber].plot for an individual appliance/meter. I would like to graphically see the results of the disaggregation of say 5 appliances compared to the ground truth plot without the clutter of other appliances.

Thanks in advance
-Shahril-

**Comment by JackKelly** [comment]
@psychdragon 

> if i wanted to plot out a selection of appliances together instead of the entire collection, how do i do that?

There are several ways to select meters.

If you want the top 5 submeters (by energy consumption) then this should work:

```python
import nilmtk
ds = nilmtk.DataSet('/data/mine/vadeec/merged/ukdale.h5')
ds.set_window("2014-01-01", "2014-02-01")
elec = ds.buildings[1].elec
top_k = elec.submeters().select_top_k(5)
top_k.plot()
```

If you want to manually specify which meters you want to plot then this should work:

```python
APPLIANCE_LIST = ["kettle", "toaster"]
meters = [elec[appliance] for appliance in APPLIANCE_LIST]
metergroup = nilmtk.MeterGroup(meters)
metergroup.plot()
```

**Comment by JackKelly** [comment]
I've just pushed some code which makes it slightly easier to select multiple meters using appliances.  Now you can do:

```python
kettle_and_toaster = elec.select_using_appliances(type=["kettle", "toaster"])
kettle_and_toaster.plot()
```

**Comment by JackKelly** [comment]
I've updated the docs.  See the section "Select subgroups of meters" in [Meter selection and basic statistics](https://github.com/nilmtk/nilmtk/blob/master/docs/manual/user_guide/elecmeter_and_metergroup.ipynb)

**Comment by nipunreddevil** [comment]
Great work Jack!

**Comment by psychdragon** [comment]
Thanks.

So, do i just do a pip upgrade to get the new code?

Sorry, basic question, i know. Just making sure.

**Comment by JackKelly** [comment]
> So, do i just do a pip upgrade to get the new code?

you'll need to do a `git pull` in your NILMTK directory.  (NILMTK isn't in PyPI yet)

**Comment by psychdragon** [comment]
Thanks @JackKelly ,
working nicely.

@nipunreddevil ,
You mentioned using the basic pandas data structure, Have been trying to convert the elec.power_series() generator data to slot into a new pandas dataframe which not much success. Any guides besides the pandas documentation i can refer to? I can't simply convert the generator to a list I have learnt(the hard way).

**Comment by psychdragon** [comment]
Answering my own question. I managed to get it done by hacking up a simple function. posting it here in case it may help others:

def convertgenerator(gen):
    buflist=[]
    for i in gen:
        buflist.append(i.values)
    return list(buflist[0])
        
totpd=pd.DataFrame({appliancelist[0]:convertgenerator(top_elec[appliancelist[0]].power_series())})
for i in appliancelist:
    print "loading "+str(i)
    
    bufpd=pd.DataFrame({i:convertgenerator(top_elec[i].power_series())})
    
    totpd = pd.concat([totpd, bufpd], axis=1)
           
totpd

**Comment by JackKelly** [comment]
> Have been trying to convert the elec.power_series() generator data to slot into a new pandas dataframe

You can convert the generator to a Pandas Series object just by doing `elec.power_series().next()`.  Or by using `elec.power_series_all_data()` (which returns a Series).

**Comment by psychdragon** [comment]
thanks jack. much appreciated.
